blockchain/utxocache.go

```diff
@@ -0,0 +1,909 @@
+			panicf("attach block node hash %v (height %v) parent hash %v does "+
+				"not match previous parent block hash %v", &n.hash, n.height,
+				&n.parent.hash, parent.Hash())
+		}
+
+		// Store the loaded block as parent of the block in the next iteration.
+		prevBlockAttached = block
+
+		// Determine if treasury agenda is active.
+		isTreasuryEnabled, err := b.isTreasuryAgendaActive(n.parent)
+		if err != nil {
+			return err
+		}
+
+		// Update the view to mark all utxos referenced by the block as
+		// spent and add all transactions being created by this block to it.
+		// In the case the block votes against the parent, also disconnect
+		// all of the regular transactions in the parent block.
+		err = view.connectBlock(b.db, block, parent, nil, isTreasuryEnabled)
+		if err != nil {
+			return err
+		}
+
+		// Commit all entries in the view to the utxo cache.  All entries in the
+		// view that are marked as modified and spent are removed from the view.
+		// Additionally, all entries that are added to the cache are removed from
+		// the view.
+		err = b.utxoCache.Commit(view)
+		if err != nil {
+			return err
+		}
+
+		// Conditionally flush the utxo cache to the database.  Don't force flush
+		// since many blocks may be connected in quick succession when initializing.
+		err = b.utxoCache.MaybeFlush(&n.hash, uint32(n.height), false, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ShutdownUtxoCache flushes the utxo cache to the database on shutdown.  Since
+// the cache is flushed periodically during initial block download and flushed
+// after every block is connected after initial block download is complete,
+// this flush that occurs during shutdown should finish relatively quickly.
+//
+// Note that if an unclean shutdown occurs, the cache will still be initialized
+// properly when restarted as during initialization it will replay blocks to
+// catch up to the tip block if it was not fully flushed before shutting down.
+// However, it is still preferred to flush when shutting down versus always
+// recovering on startup since it is faster.
+//
+// This function should only be called during shutdown.
+func (b *BlockChain) ShutdownUtxoCache() {
+	b.chainLock.RLock()
+	defer b.chainLock.RUnlock()
+
+	tip := b.bestChain.Tip()
+
+	// Force a cache flush and log the flush details.
+	b.utxoCache.MaybeFlush(&tip.hash, uint32(tip.height), true, true)
+}
+
+// FetchUtxoEntry loads and returns the requested unspent transaction output
+// from the point of view of the the main chain tip.
+//
+// NOTE: Requesting an output for which there is no data will NOT return an
+// error.  Instead both the entry and the error will be nil.  This is done to
+// allow pruning of spent transaction outputs.  In practice this means the
+// caller must check if the returned entry is nil before invoking methods on it.
+//
+// This function is safe for concurrent access however the returned entry (if
+// any) is NOT.
```

Comment:
Because of `entry.Clone()` isn't the entry now safe to mutate? Or this comment has a different meaning?

Reply:
No, it's still not safe because multiple goroutines might be accessing the same cloned entry.

Reply:
Right, it is safe to be mutated, but it is still not safe for concurrent access, as Dave stated.